File: cafe-management/Frontend/src/app/services/menu.service.ts
import { HttpClient, HttpHeaders, HttpParams } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { environment } from 'src/environments/environment';

/**
 * Service for Menu API operations
 * Handles HTTP requests for fetching menu data
 */
@Injectable({
  providedIn: 'root'
})
export class MenuService {

  url = environment.apiUrl;

  constructor(private httpClient: HttpClient) { }

  /**
   * Get complete menu with all categories and items
   * @returns Observable with complete menu response
   */
  getMenu(): Observable<any> {
    return this.httpClient.get(this.url + "/menu/get");
  }

  /**
   * Get menu items filtered by category
   * @param categoryId Category ID to filter by
   * @returns Observable with filtered menu response
   */
  getMenuByCategory(categoryId: number): Observable<any> {
    const params = new HttpParams().set('categoryId', categoryId.toString());
    return this.httpClient.get(this.url + "/menu/getByCategory", { params });
  }

  /**
   * Get paginated menu
   * @param page Page number (0-indexed)
   * @param size Number of items per page
   * @returns Observable with paginated menu response
   */
  getMenuPaginated(page: number = 0, size: number = 20): Observable<any> {
    const params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString());
    return this.httpClient.get(this.url + "/menu/getPaginated", { params });
  }

  /**
   * Generate menu PDF
   * @param showVegOnly Filter to show only vegetarian items
   * @returns Observable with PDF as blob
   */
  generateMenuPdf(showVegOnly: boolean = false): Observable<Blob> {
    const params = new HttpParams().set('showVegOnly', showVegOnly.toString());
    return this.httpClient.get(this.url + "/menu/generatePdf", { 
      params,
      responseType: 'blob'
    });
  }
}

File: cafe-management/com.inn.cafe/src/main/java/com/inn/cafe/service/MenuService.java
package com.inn.cafe.service;

import com.inn.cafe.wrapper.MenuResponseWrapper;
import org.springframework.http.ResponseEntity;

/**
 * Service interface for Menu business logic
 */
public interface MenuService {

    /**
     * Get complete menu with all categories and items
     * Items are grouped by category and sorted appropriately
     * 
     * @return ResponseEntity with complete menu data
     */
    ResponseEntity<MenuResponseWrapper> getMenu();

    /**
     * Get menu items for a specific category
     * 
     * @param categoryId Category ID to filter by
     * @return ResponseEntity with menu data for the category
     */
    ResponseEntity<MenuResponseWrapper> getMenuByCategory(Integer categoryId);

    /**
     * Get paginated menu items
     * Useful for large menus or mobile views
     * 
     * @param page Page number (0-indexed)
     * @param size Number of items per page
     * @return ResponseEntity with paginated menu data
     */
    ResponseEntity<MenuResponseWrapper> getMenuPaginated(Integer page, Integer size);

    /**
     * Generate menu PDF
     * 
     * @param showVegOnly Filter to show only vegetarian items
     * @return ResponseEntity containing PDF as byte array
     */
    ResponseEntity<byte[]> generateMenuPdf(Boolean showVegOnly);
}
